pyroomacoustics/open_sofa_interpolate.py

```diff
@@ -0,0 +1,556 @@
+            R_z = np.array([[np.cos(n_a), -np.sin(n_a), 0], [np.sin(n_a), np.cos(n_a), 0], [0, 0, 1]])
+            res = np.matmul(R_z, R_y)
+            '''
+
+
+            cart_points=np.empty((3,self.phi_sofa.shape[0]))
+            cart_points[0,:]=self.sofa_x
+            cart_points[1,:]=self.sofa_y
+            cart_points[2,:]=self.sofa_z
+
+            self.rotated_sofa_points = np.matmul(res, cart_points)
+
+            self.rotated_sofa_phi, self.rotated_sofa_theta, self.rotated_sofa_r = cart2sphere(self.rotated_sofa_points.T)
+
+            print(np.max(self.rotated_sofa_phi),np.min(self.rotated_sofa_phi))
+            self.nn_kd_tree_rotated_sofa_grid = cKDTree(
+                np.hstack((self.rotated_sofa_phi.reshape(-1, 1), self.rotated_sofa_theta.reshape(-1, 1))))
+
+
+
+    def cal_spherical_coeffs_grid_and_interpolate(self):
+
+        """
+        Weighted least square solution to calculate discrete sphreical harmonic coeffs , Using the spherical harmonic coeffs
+        interpolation is done on fibonacci sphere
+        """
+
+
+        #freq_angles_fft = np.zeros((self.no_of_nodes, self.samples_size_ir// 2), dtype=np.complex_)
+
+        # Take n-point dft of the FIR's on the grid
+        for i in range(self.no_of_nodes):
+            self.freq_angles_fft[i, :] = fft(self.sofa_msr_fir[i, :])  # [:(self.samples_size_ir // 2)] #self.samples_size_ir//2
+
+
+        g_tilda = np.matmul(self.w_, self.freq_angles_fft)
+
+        #Shape w_ : (540*540) or (480*480)
+        #Shape g_tilda : (480*256),(480*128),(540*2048),(540*1048),(540,683)
+
+        gamma_full_scale = np.matmul(self.Ysh_tilda_inv, g_tilda) #Coeffs for every freq band , for all the nodes present in the sphere
+
+        '''
+        #Shape gamma_full_scale : ((deg+1)^2 * 256 | 128 | 2048 | 1024 | 683) Coeffs for every frequency bin
+
+        #Select coeffs for particular frequency for plotting purpose
+        #For sources select between (0-1023), receivers (0-128)
+
+        #freq_bin=20
+        #gamma = gamma_full_scale[:, freq_bin] #[:(self.samples_size_ir//2)]
+
+        #Shape gamma : ((deg+1)^2) Coeffs for a particular frequency bin .
+        '''
+
+        self.sh_coeffs_expanded_original_grid = np.fft.ifft(np.matmul(self.Ysh, gamma_full_scale),axis=-1)
+        self.sh_coeffs_expanded_target_grid = np.fft.ifft(np.matmul(self.Ysh_fibo, gamma_full_scale),axis=-1)
+        print(self.sh_coeffs_expanded_target_grid.shape,self.sh_coeffs_expanded_target_grid.shape,self.samples_size_ir)
+
+
+
+    def open_sofa_database(self, path, fs=16000):
+            # Open DirPat database
+
+            if self.source == True:
+                file_sofa = sofa.Database.open(path)
+
+                # Receiver positions
+
+                rcv_pos = file_sofa.Receiver.Position.get_values()
+
+                print(rcv_pos.shape)
+
+                # CHEAP HACCCKK SPECIFICALLY FOR DIRPAT
```

Comment:
Will this fail if we use this code for general sofa files ? How difficult would it be to make it generic ?